### PR TITLE
Add search engine and quick search for repositories

### PR DESCRIPTION
**Pull request by sdorra**
## Proposed changes

Add a powerful search engine based on lucene to the scm-manager api.
The api can be used to index objects, simply by annotating them and add them to an index.
The first indexed object is the repository which could queried by quick search in the header.

This pr is one of a series to add a omni search experience to scm-manager. Documentation follows if the whole search becomes stable.

### Your checklist for this pull request

**Contributor**:
- [X] PR is well described and the description can be used as a commit message on squash
- [X] Related issues linked to PR if existing and labels set
- [X] Target branch is not master (in most cases develop should bet the target of choice) 
- [X] Code does not conflict with target branch
- [X] New code is covered with unit tests
- [X] [Changelog entry file](https://github.com/scm-manager/changelog#changelog-entry-files) created in `gradle/changelog` or CHANGELOG.md is updated for plugins

**Reviewer**:
- [ ] The clean code principles are respected ([CleanCode](https://clean-code-developer.com/virtues/))
- [ ] All new code/logic is implemented on the right spot / "Should this be done here?"
- [ ] UI changes fits into the layout
- [ ] The UI views / components are responsive (mobile views)
- [ ] Correct translations are available

### Checklist for branch merge request (not required for forks)

- [ ] Branch is green/blue on [Jenkins](https://oss.cloudogu.com/jenkins/)
- [ ] Quality Gate passed on [SonarQube](https://sonarcloud.io/organizations/scm-manager/projects)

### CODE REVIEW

**Review by pfeuffer — COMMENTED** [review]
Sometimes I get a repository as a result more than once. This seems odd to me.
image

File: scm-webapp/src/main/java/sonia/scm/search/UnsupportedTypeOfFieldException.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package sonia.scm.search;
+
+public class UnsupportedTypeOfFieldException extends SearchEngineException {
+  public UnsupportedTypeOfFieldException(String message) {
+    super(message);
+  }
```
Comment (pfeuffer) [inline scm-webapp/src/main/java/sonia/scm/search/UnsupportedTypeOfFieldException.java:30]: Again, super minor, but I would have liked a constructor like this more:
```suggestion
  public UnsupportedTypeOfFieldException(Class<?> fieldType, Field field) {
    super("type " + fieldType + " of " + field.getName() + " is unsupported.");
  }
```

File: scm-webapp/src/main/java/sonia/scm/search/ValueExtractors.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+ */
+
+package sonia.scm.search;
+
+import org.apache.lucene.index.IndexableField;
+
+import javax.annotation.Nonnull;
+import java.time.Instant;
+
+class ValueExtractors {
```
Comment (pfeuffer) [inline scm-webapp/src/main/java/sonia/scm/search/ValueExtractors.java:32]:
```suggestion
final class ValueExtractors {
```

File: scm-webapp/src/test/java/sonia/scm/api/v2/resources/SearchResourceTest.java [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.lenient;
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class)
+class SearchResourceTest {
```
Comment (pfeuffer) [inline scm-webapp/src/test/java/sonia/scm/api/v2/resources/SearchResourceTest.java:67]: A good class to advertise the relatively new class `JsonMockHttpResponse` ;-)

File: scm-webapp/src/test/java/sonia/scm/repository/IndexUpdateListenerTest.java [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+  @Mock
+  private IndexLogStore indexLogStore;
+
+  @InjectMocks
+  private IndexUpdateListener updateListener;
+
+  @Test
+  @SuppressWarnings("java:S6068")
+    // eq is required for verify with changing parametern
```
Comment (pfeuffer) [inline scm-webapp/src/test/java/sonia/scm/repository/IndexUpdateListenerTest.java:77]: No, not really. At least the test is green without `eq`, too

**Review by pfeuffer — APPROVED** [review]
Minor code suggestions. No real issues.

**Comment by pfeuffer** [comment]
What about `@Beta` annotations for the search core api?

File: scm-core/src/main/java/sonia/scm/search/SearchEngine.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package sonia.scm.search;
```
[the remark this answers is not in the document]
Comment (sdorra) [reply scm-core/src/main/java/sonia/scm/search/SearchEngine.java:25]: I've marked all exposed search api classes as beta

File: scm-core/src/main/java/sonia/scm/search/NoDefaultQueryFieldsFoundException.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+/**
+ * Exception is thrown if a best guess query is executed
+ * and the search type does not contain a field which is marked for default query.
+ *
+ * @see Indexed#defaultQuery()
+ * @since 2.21.0
+ */
+public class NoDefaultQueryFieldsFoundException extends SearchEngineException {
+  public NoDefaultQueryFieldsFoundException(String message) {
```
[the remark this answers is not in the document]
Comment (sdorra) [reply scm-core/src/main/java/sonia/scm/search/NoDefaultQueryFieldsFoundException.java:38]: Done

**Pull request merged**